File: app/src/main/java/it/cammino/risuscito/LUtils.java
package it.cammino.risuscito;

import android.annotation.TargetApi;
import android.app.Activity;
import android.content.Intent;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.Build;
import android.preference.PreferenceManager;
import android.support.annotation.NonNull;
import android.support.v4.content.FileProvider;
import android.support.v4.view.ViewCompat;
import android.support.v4.view.animation.FastOutSlowInInterpolator;
import android.text.Html;
import android.text.Spanned;
import android.util.Log;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.view.animation.Interpolator;

import com.google.firebase.crash.FirebaseCrash;

import org.w3c.dom.Document;
import org.w3c.dom.Element;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.StringWriter;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import javax.xml.transform.Transformer;
import javax.xml.transform.TransformerConfigurationException;
import javax.xml.transform.TransformerException;
import javax.xml.transform.TransformerFactory;
import javax.xml.transform.dom.DOMSource;
import javax.xml.transform.stream.StreamResult;

import it.cammino.risuscito.database.RisuscitoDatabase;
import it.cammino.risuscito.database.dao.CronologiaDao;
import it.cammino.risuscito.database.entities.Cronologia;

public class LUtils {

    final String TAG = getClass().getCanonicalName();

    private static final Interpolator INTERPOLATOR = new FastOutSlowInInterpolator();

    private final static String FILE_FORMAT = ".risuscito";

    private Activity mActivity;

    private LUtils(Activity activity) {
        mActivity = activity;
    }

    public static LUtils getInstance(Activity activity) {
        return new LUtils(activity);
    }

    public static boolean hasL() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP;
    }

    public static boolean hasO() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.O;
    }

    public void startActivityWithTransition(final Intent intent, final View clickedView,
                                            final String transitionName) {
//        ActivityOptions options = null;
//        if (hasL() && clickedView != null && !TextUtils.isEmpty(transitionName)) {
//            options = ActivityOptions.makeSceneTransitionAnimation(
//                    mActivity, clickedView, transitionName);
//            ActivityCompat.startActivity(mActivity, intent, options.toBundle());
//        } else {
        mActivity.startActivity(intent);
        mActivity.overridePendingTransition(R.anim.slide_in_right, R.anim.hold_on);
//        }

        //aggiorno la cronologia
//        DatabaseCanti listaCanti = new DatabaseCanti(mActivity);
//        SQLiteDatabase db = listaCanti.getReadableDatabase();
//
//        String select = "SELECT COUNT(*) FROM CRONOLOGIA" +
//                "        WHERE id_canto = " + intent.getExtras().getInt("idCanto");
//        Cursor lista = db.rawQuery(select, null);
//        lista.moveToFirst();
//
//        switch (lista.getInt(0)) {
//            case 1:
//                String update = "UPDATE CRONOLOGIA" +
//                        " SET ultima_visita = CURRENT_TIMESTAMP" +
//                        " WHERE id_canto = " + intent.getExtras().getInt("idCanto");
//                db.execSQL(update);
//                break;
//            case 0:
//                ContentValues values = new ContentValues();
//                values.put("id_canto", intent.getExtras().getInt("idCanto"));
//                db.insert("CRONOLOGIA", null, values);
//                break;
//        }
//
//        lista.close();
//        db.close();

        new Thread(new Runnable() {
            @Override
            public void run() {
                CronologiaDao mDao =
                        RisuscitoDatabase.getInstance(mActivity).cronologiaDao();
                Cronologia cronologia = new Cronologia();
                cronologia.idCanto =  intent.getExtras().getInt("idCanto");
                mDao.insertCronologia(cronologia);
            }
        }).start();

    }

//    public void startActivityWithTransition(Intent intent) {
//        mActivity.startActivity(intent);
//        mActivity.overridePendingTransition(R.anim.slide_in_right, R.anim.hold_on);
//    }


//    public void startActivityWithFadeIn(Intent intent, final View clickedView,
//                                        final String transitionName) {
////        ActivityOptions options = null;
////        if (hasL() && clickedView != null && !TextUtils.isEmpty(transitionName)) {
////            options = ActivityOptions.makeSceneTransitionAnimation(
////                    mActivity, clickedView, transitionName);
////            ActivityCompat.startActivity(mActivity, intent, options.toBundle());
////        } else {
//        mActivity.startActivity(intent);
//        mActivity.overridePendingTransition(R.anim.image_fade_in, R.anim.hold_on);
////        }
//    }

    void startActivityWithFadeIn(Intent intent) {
        mActivity.startActivity(intent);
        mActivity.overridePendingTransition(R.anim.image_fade_in, R.anim.hold_on);
    }

    void closeActivityWithTransition() {
//        if (hasL())
//            mActivity.finishAfterTransition();
//        else {
        mActivity.finish();
        mActivity.overridePendingTransition(0, R.anim.slide_out_right);
//        }
    }

    void closeActivityWithFadeOut() {
//        if (hasL())
//            mActivity.finishAfterTransition();
//        else {
        mActivity.finish();
        mActivity.overridePendingTransition(0, R.anim.image_fade_out);
//        }
    }

    void goFullscreen() {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.KITKAT) {
            mActivity.requestWindowFeature(Window.FEATURE_NO_TITLE);
            mActivity.getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN);
        } else mActivity.getWindow().getDecorView().setSystemUiVisibility(
                View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                        | View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
                        | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                        | View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
                        | View.SYSTEM_UI_FLAG_FULLSCREEN
                        | View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY);
    }


//    public void applyFontedTab(ViewPager viewPager, TabLayout tabLayout) {
//        TabLayout.Tab mTab;
//        for (int i = 0; i < viewPager.getAdapter().getCount(); i++) {
//            TextView tv = (TextView) mActivity.getLayoutInflater().inflate(R.layout.item_tab, null);
//            if (i == viewPager.getCurrentItem()) tv.setSelected(true);
//            tv.setText(viewPager.getAdapter().getPageTitle(i));
//            mTab = tabLayout.getTabAt(i);
//            if (mTab != null)
//                mTab.setCustomView(tv);
//        }
//    }

//    public static boolean hasICS() {
//        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.ICE_CREAM_SANDWICH;
//    }

    public static boolean hasJB() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1;
    }

    public static boolean hasN() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.N;
    }

    Uri listToXML(@NonNull ListaPersonalizzata lista) {

        try {
            DocumentBuilderFactory docFactory = DocumentBuilderFactory.newInstance();
            DocumentBuilder docBuilder = docFactory.newDocumentBuilder();

            // root elements
            Document doc = docBuilder.newDocument();
            Element rootElement = doc.createElement("list");
            rootElement.setAttribute("title", lista.getName());
            doc.appendChild(rootElement);

            for (int i = 0; i < lista.getNumPosizioni(); i++) {
                Element position = doc.createElement("position");
                position.setAttribute("name", lista.getNomePosizione(i));
                if (!lista.getCantoPosizione(i).equals(""))
                    position.appendChild(doc.createTextNode(lista.getCantoPosizione(i)));
                else
                    position.appendChild(doc.createTextNode("0"));
                rootElement.appendChild(position);
            }

            DOMSource domSource = new DOMSource(doc);
            StringWriter writer = new StringWriter();
            StreamResult result = new StreamResult(writer);
            TransformerFactory tf = TransformerFactory.newInstance();
            Transformer transformer = tf.newTransformer();
            transformer.transform(domSource, result);
            Log.d(TAG, "listToXML: " + writer.toString());
//            writer.toString();

            File exportFile = new File(mActivity.getCacheDir().getAbsolutePath() + "/" + lista.getName() + FILE_FORMAT);
            Log.d(TAG, "listToXML: exportFile = " + exportFile.getAbsolutePath());
            FileOutputStream fos = new FileOutputStream(exportFile);
            String dataWrite = writer.toString();
            fos.write(dataWrite.getBytes());
            fos.close();

            return FileProvider.getUriForFile(mActivity, "it.cammino.risuscito.fileprovider", exportFile);

        }
        catch (ParserConfigurationException e) {
            Log.e(TAG, "listToXML: " + e.getLocalizedMessage(), e);
            FirebaseCrash.log(e.getMessage());
            return null;
        }
        catch (TransformerConfigurationException e) {
            Log.e(TAG, "listToXML: " + e.getLocalizedMessage(), e);
            FirebaseCrash.log(e.getMessage());
            return null;
        }
        catch (TransformerException e) {
            Log.e(TAG, "listToXML: " + e.getLocalizedMessage(), e);
            FirebaseCrash.log(e.getMessage());
            return null;
        }
        catch (FileNotFoundException e) {
            Log.e(TAG, "listToXML: " + e.getLocalizedMessage(), e);
            FirebaseCrash.log(e.getMessage());
            return null;
        }
        catch (IOException e) {
            Log.e(TAG, "listToXML: " + e.getLocalizedMessage(), e);
            FirebaseCrash.log(e.getMessage());
            return null;
        }

    }

    public boolean isOnTablet() {
        return (mActivity.getResources().getBoolean(R.bool.is_tablet));

    }

    public void convertIntPreferences() {
        convert(Utility.DEFAULT_INDEX);
        convert(Utility.SAVE_LOCATION);
    }

    private void convert(String prefName) {
        SharedPreferences pref =  PreferenceManager.getDefaultSharedPreferences(mActivity);
        try {
            pref.getString(prefName, "0");
            Log.d(TAG, "onCreateView: " + prefName + " STRING");
        }
        catch (ClassCastException e) {
            Log.d(TAG, "onCreateView: " + prefName + " INTEGER >> CONVERTO");
            SharedPreferences.Editor editor = PreferenceManager
                    .getDefaultSharedPreferences(mActivity)
                    .edit();
            editor.putString(prefName, String.valueOf(pref.getInt(prefName, 0)));
            editor.apply();
        }
    }

    // Same animation that FloatingActionButton.Behavior uses to show the FAB when the AppBarLayout enters
    void animateIn(View view) {
//        if (view.getVisibility() == View.INVISIBLE) {
        view.setVisibility(View.VISIBLE);
        ViewCompat.animate(view)
                .setDuration(200)
                .translationY(0)
                .setInterpolator(INTERPOLATOR).withLayer().setListener(null)
                .start();
//        }
    }

    @SuppressWarnings("deprecation")
    private static Spanned fromHtmlLegacy(String input) {
        return Html.fromHtml(input);
    }

    @TargetApi(Build.VERSION_CODES.N)
    private static Spanned fromHtml(String input) {
        return Html.fromHtml(input, Html.FROM_HTML_MODE_LEGACY);
    }

    public static Spanned fromHtmlWrapper(String input) {
        if (LUtils.hasN())
            return fromHtml(input);
        else
            return fromHtmlLegacy(input);
    }

}

File: app/src/main/java/it/cammino/risuscito/LUtils.kt
package it.cammino.risuscito

import android.Manifest
import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.animation.TimeInterpolator
import android.annotation.TargetApi
import android.app.Activity
import android.app.ActivityOptions
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import android.text.Html
import android.text.Spanned
import android.util.Log
import android.view.View
import android.view.View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR
import android.view.WindowManager
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.core.content.FileProvider
import androidx.core.content.edit
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.WindowInsetsControllerCompat
import androidx.core.view.isVisible
import androidx.interpolator.view.animation.FastOutLinearInInterpolator
import androidx.interpolator.view.animation.LinearOutSlowInInterpolator
import androidx.lifecycle.lifecycleScope
import androidx.preference.PreferenceManager
import com.google.firebase.crashlytics.FirebaseCrashlytics
import it.cammino.risuscito.database.RisuscitoDatabase
import it.cammino.risuscito.database.entities.Cronologia
import it.cammino.risuscito.ui.Animations
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.io.*
import javax.xml.parsers.DocumentBuilderFactory
import javax.xml.parsers.ParserConfigurationException
import javax.xml.transform.TransformerConfigurationException
import javax.xml.transform.TransformerException
import javax.xml.transform.TransformerFactory
import javax.xml.transform.dom.DOMSource
import javax.xml.transform.stream.StreamResult

class LUtils private constructor(private val mActivity: Activity) {

    val isOnTablet: Boolean
        get() = mActivity.resources.getBoolean(R.bool.is_tablet)

    val hasThreeColumns: Boolean
        get() = mActivity.resources.getBoolean(R.bool.has_three_columns)

    val isGridLayout: Boolean
        get() = mActivity.resources.getBoolean(R.bool.is_grid_layout)

    val isLandscape: Boolean
        get() = mActivity.resources.getBoolean(R.bool.landscape)

    val isFabExpansionLeft: Boolean
        get() = mActivity.resources.getBoolean(R.bool.fab_orientation_left)

    fun startActivityWithTransition(
        intent: Intent,
        startView: View?
    ) {

        val options = ActivityOptions.makeSceneTransitionAnimation(
            mActivity,
            startView,
            "shared_element_container" // The transition name to be matched in Activity B.
        )

        mActivity.startActivity(intent, options.toBundle())

        val mDao = RisuscitoDatabase.getInstance(mActivity).cronologiaDao()
        val cronologia = Cronologia()
        cronologia.idCanto = intent.extras?.getInt(Utility.ID_CANTO) ?: 0
        (mActivity as? AppCompatActivity)?.lifecycleScope?.launch(Dispatchers.IO) {
            mDao.insertCronologia(
                cronologia
            )
        }
    }

    fun startActivityWithFadeIn(intent: Intent) {
        mActivity.startActivity(intent)
        Animations.enterZoom(mActivity)
    }

    //ISSUE in API 21
    fun finishAfterTransitionWrapper() {
        if (hasM())
            mActivity.finishAfterTransition()
        else
            mActivity.finish()
    }

    fun setLigthStatusBar(light: Boolean) {
        WindowInsetsControllerCompat(
            mActivity.window,
            mActivity.window.decorView
        ).isAppearanceLightStatusBars = light
        setLighStatusBarFlag(light)
    }

    private fun setLighStatusBarFlag(light: Boolean) {
        if (hasM())
            setLighStatusBarFlagM(light)
    }

    @Suppress("DEPRECATION")
    @RequiresApi(Build.VERSION_CODES.M)
    private fun setLighStatusBarFlagM(light: Boolean) {
        if (light)
            mActivity
                .window
                .decorView.systemUiVisibility = SYSTEM_UI_FLAG_LIGHT_STATUS_BAR
    }

    internal fun goFullscreen() {
        val windowInsetsController =
            ViewCompat.getWindowInsetsController(mActivity.window.decorView) ?: return
        // Configure the behavior of the hidden system bars
        windowInsetsController.systemBarsBehavior =
            WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
        // Hide both the status bar and the navigation bar
        windowInsetsController.hide(WindowInsetsCompat.Type.systemBars())
//        when {
//            hasR() -> goFullscreenR()
//            else -> goFullscreenLegacy()
//        }
    }

//    private fun goFullscreenR() {
//        WindowCompat.setDecorFitsSystemWindows(mActivity.window, false)
//        WindowInsetsControllerCompat(
//            mActivity.window,
//            mActivity.window.decorView
//        ).let { controller ->
//            controller.hide(WindowInsetsCompat.Type.statusBars() or WindowInsetsCompat.Type.navigationBars())
//            controller.systemBarsBehavior =
//                WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
//        }
//    }

//    @Suppress("DEPRECATION")
//    private fun goFullscreenLegacy() {
//        mActivity
//            .window
//            .decorView.systemUiVisibility = (View.SYSTEM_UI_FLAG_LAYOUT_STABLE
//                or View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
//                or View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
//                or View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
//                or View.SYSTEM_UI_FLAG_FULLSCREEN
//                or View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY)
//    }

    // controlla se l'app deve mantenere lo schermo acceso
    fun checkScreenAwake() {
        val pref = PreferenceManager.getDefaultSharedPreferences(mActivity)
        val screenOn = pref.getBoolean(Utility.SCREEN_ON, false)
        if (screenOn)
            mActivity.window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
        else
            mActivity.window.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
    }

    internal fun listToXML(lista: ListaPersonalizzata?): Uri? {
        lista?.let {
            try {
                val docFactory = DocumentBuilderFactory.newInstance()
                val docBuilder = docFactory.newDocumentBuilder()

                // root elements
                val doc = docBuilder.newDocument()
                val rootElement = doc.createElement("list")
                rootElement.setAttribute("title", it.name)
                doc.appendChild(rootElement)

                for (i in 0 until it.numPosizioni) {
                    val position = doc.createElement("position")
                    position.setAttribute("name", it.getNomePosizione(i))
                    if (it.getCantoPosizione(i) != "")
                        position.appendChild(doc.createTextNode(it.getCantoPosizione(i)))
                    else
                        position.appendChild(doc.createTextNode("0"))
                    rootElement.appendChild(position)
                }

                val domSource = DOMSource(doc)
                val writer = StringWriter()
                val result = StreamResult(writer)
                val tf = TransformerFactory.newInstance()
                val transformer = tf.newTransformer()
                transformer.transform(domSource, result)
                Log.d(TAG, "listToXML: $writer")

                val exportFile = File(mActivity.cacheDir.absolutePath + "/" + it.name + FILE_FORMAT)
                Log.d(TAG, "listToXML: exportFile = " + exportFile.absolutePath)
                val fos = FileOutputStream(exportFile)
                val dataWrite = writer.toString()
                fos.write(dataWrite.toByteArray())
                fos.close()

                return FileProvider.getUriForFile(
                    mActivity,
                    "it.cammino.risuscito.fileprovider",
                    exportFile
                )

            } catch (e: ParserConfigurationException) {
                Log.e(TAG, "listToXML: " + e.localizedMessage, e)
                FirebaseCrashlytics.getInstance().recordException(e)
                return null
            } catch (e: TransformerConfigurationException) {
                Log.e(TAG, "listToXML: " + e.localizedMessage, e)
                FirebaseCrashlytics.getInstance().recordException(e)
                return null
            } catch (e: TransformerException) {
                Log.e(TAG, "listToXML: " + e.localizedMessage, e)
                FirebaseCrashlytics.getInstance().recordException(e)
                return null
            } catch (e: FileNotFoundException) {
                Log.e(TAG, "listToXML: " + e.localizedMessage, e)
                FirebaseCrashlytics.getInstance().recordException(e)
                return null
            } catch (e: IOException) {
                Log.e(TAG, "listToXML: " + e.localizedMessage, e)
                FirebaseCrashlytics.getInstance().recordException(e)
                return null
            }
        }
        Log.e(TAG, "input lista null")
        return null
    }

    fun convertIntPreferences() {
        convert(Utility.DEFAULT_INDEX)
        convert(Utility.SAVE_LOCATION)
        convert(Utility.DEFAULT_SEARCH)
    }

    private fun convert(prefName: String) {
        val pref = PreferenceManager.getDefaultSharedPreferences(mActivity)
        try {
            pref.getString(prefName, "0")
            Log.d(TAG, "onCreateView: $prefName STRING")
        } catch (e: ClassCastException) {
            Log.d(TAG, "onCreateView: $prefName INTEGER >> CONVERTO")
            PreferenceManager.getDefaultSharedPreferences(mActivity)
                .edit { putString(prefName, pref.getInt(prefName, 0).toString()) }
        }

    }

    // Same animation that FloatingActionButton.Behavior uses to show the FAB when the AppBarLayout
    // enters
    internal fun animateIn(view: View) {
        view.isVisible = true
        view.animate().translationY(0f)
            .setInterpolator(LINEAR_OUT_SLOW_IN_INTERPOLATOR).setDuration(225L)
            .setListener(null).start()
    }

    internal fun animateOut(view: View) {
        view.animate().translationY(view.height.toFloat())
            .setInterpolator(FAST_OUT_LINEAR_IN_INTERPOLATOR).setDuration(175L)
            .setListener(
                object : AnimatorListenerAdapter() {
                    override fun onAnimationEnd(animation: Animator?) {
                        view.isVisible = false
                    }
                }
            ).start()
    }

    val hasStorageAccess: Boolean
        get() = hasQ() || ContextCompat.checkSelfPermission(
            mActivity,
            Manifest.permission.WRITE_EXTERNAL_STORAGE
        ) == PackageManager.PERMISSION_GRANTED

    companion object {

        private const val FILE_FORMAT = ".risuscito"
        internal val TAG = LUtils::class.java.canonicalName
        internal val FAST_OUT_LINEAR_IN_INTERPOLATOR: TimeInterpolator =
            FastOutLinearInInterpolator()
        internal val LINEAR_OUT_SLOW_IN_INTERPOLATOR: TimeInterpolator =
            LinearOutSlowInInterpolator()

        fun getInstance(activity: Activity): LUtils {
            return LUtils(activity)
        }

        fun hasO(): Boolean {
            return Build.VERSION.SDK_INT >= Build.VERSION_CODES.O
        }

        fun hasQ(): Boolean {
            return Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q
        }

        fun hasM(): Boolean {
            return Build.VERSION.SDK_INT >= Build.VERSION_CODES.M
        }

        fun hasN(): Boolean {
            return Build.VERSION.SDK_INT >= Build.VERSION_CODES.N
        }

        fun hasP(): Boolean {
            return Build.VERSION.SDK_INT >= Build.VERSION_CODES.P
        }

        fun hasR(): Boolean {
            return Build.VERSION.SDK_INT >= Build.VERSION_CODES.R
        }

        fun hasS(): Boolean {
            return Build.VERSION.SDK_INT >= Build.VERSION_CODES.S
        }

        @Suppress("DEPRECATION")
        private fun fromHtmlLegacy(input: String): Spanned {
            return Html.fromHtml(input)
        }

        @TargetApi(Build.VERSION_CODES.N)
        private fun fromHtml(input: String): Spanned {
            return Html.fromHtml(input, Html.FROM_HTML_MODE_LEGACY)
        }

        fun fromHtmlWrapper(input: String): Spanned {
            return if (hasN())
                fromHtml(input)
            else
                fromHtmlLegacy(input)
        }

        fun getResId(resName: String?, c: Class<*>): Int {
            resName?.let {
                return try {
                    val idField = c.getDeclaredField(it)
                    idField.getInt(idField)
                } catch (e: Exception) {
                    Log.e(TAG, "getResId: $resName" + e.localizedMessage, e)
                    -1
                }
            }
            Log.e(TAG, "resName NULL")
            return -1
        }

    }

}
